### DISCUSSION TITLE
ERR_RTE (Routing problem) when connecting only TX

### DISCUSSION THREAD

**Discussion by biemster** [holysnippet/pico_eth#5] [Q&A]
Hi @holysnippet!
Not to open an issue on this, because I think the problem is on my side. When I connect only the TX lines to the pico, and try a UDP firehose/beacon example like this: [Pico UDP beacon](https://github.com/raspberrypi/pico-examples/pull/275/files#diff-6e4120492f60050bb7ecab63a38e90b340fb477a155c33464cbf572cfa5fc782), I keep getting an error -4, which is RTE (routing problem), probably because the stack did not finish setting up.
I'm using `config.enable_dhcp_client = false;` and a static IP address.
Does your implementation needs to receive an NLP on RX, which could cause this?

**Comment by holysnippet** [comment]
Hello, I don't know what can happen. Currently I don't take into account NLP (I don't have enough instructions in the PIO). I wonder if the congestion would not be at the ARP level instead. You can try to disable ARP in lwipopts.h, otherwise I don't know!

**Reply by biemster** [reply]
I just tried with the RX connected as well so it should be able to ARP, and it still got ERR_RTE :(
I'll now first try with your demo app to see if I connected everything properly, but my desk is a mess and my network even more so it will take a couple of days..

**Reply by holysnippet** [reply]
Interesting, keep me posted. The algorithm can't correct the RX inversion, check the polarity just in case. Take the biasing voltage measurement, you should have 0.45 * the 3.3V voltage of the Pico. I'm not very efficient, I should add an Ethernet activity led. Don't hesitate to give me the number of your Ethernet transformer, I'll see if it is special.

**Reply by biemster** [reply]
So I just got around trying the demo app, on a waveshare RP2040-zero. The switch does not pick up the connection at all, so I got something fundamental wrong :(.
I've connected TX- to pin 9 (and TX+ to 10), because the Zero has an RGB LED connected to 16 (and it also fitted better). I made the appropriate change to `config.tx_neg_pin`, but the RGB is still changing color, indicating pin 16 is keeps being toggled! This might be completely unrelated, but I'm going to start investigating there. Although it is much more likely my passives are wrong, but I'm not very skilled in debugging that.

**Reply by biemster** [reply]
I've also noticed your mention of C3 and C4 on the TX lines, to block the DC component of the signal at the pico output. I have of course left them out :man_facepalming:  because you wrote it _might_ work without them as well. I don't know what type of transformer is in my switch (huawei b315-22 4g router), but this could very well be the cause of my switch not recognizing the pico.

**Reply by biemster** [reply]
O darn disregard all that, I'm working in two projects at the same time and did not change `config.tx_neg_ping` on the one I actually flashed... :no_mouth:

**Reply by holysnippet** [reply]
They are not normally required because many transformers do not have the windings connected. I can't find the reference of the transformer from the model and brand of your switch.

Regarding the LED it's weird, I tested the pin changes because my SDR uses 16 and 17 for the I2S of the ADC. I checked the source code I don't understand because nothing is hard coded.

Feel free to ask me if you think I can help you. I've been using this setup and code for almost a month intensively (streaming audio to and from my SDRs). It does the job here so much that I forgot about it! I managed to free a DMA channel in the beta branch. It has been working for a few days with the same performance. I will soon pass the code (as is because it works) into the main branch.

**Reply by biemster** [reply]
yeah the issue was totally on my side. With the TX pin correctly set I can see my router now bringing the port up, and the dhcp server handing out a lease! It does not work all the time though, and the pico does not seem to receive anything (pinging it does not work), so I have to work on the passives a bit more probably.

**Reply by holysnippet** [reply]
Good news, for the RX I advise you to be as close as possible to the values in the diagram.

The biasing voltage is critical! Resistors R1 and R2 must be exact!

Also make sure your system clock is at 120MHz (or 240 but try 120 to start with) otherwise the reception will not work.

Thanks for the feedback!

**Reply by biemster** [reply]
darn! I did not have a 2k7 for R2, and figured a 2k2 is close enough.. gives a division factor of 0.4 instead of 0.45. Since @cnlohr is not using a voltage divider bridge at all:
https://github.com/cnlohr/espthernet/blob/gh-pages/hardware/nodemcu-ethernet.png
I was hoping I would get away with it.
To be honest, I find it quite amazing that both your and his diagram both work while seemingly be completely different! But then again, my analog knowledge does not go much further than following schematics I find on the net. Do you understand what is going on on his circuit?

**Reply by holysnippet** [reply]
You can try to find a 10k ohm potentiometer to replace R1 and R2. You will have to adjust it very well!

I don't understand his schematic at all, I had seen it before. It seems to use a different type of coupling and bias. This schematic will not work at all with the Pico.

**Reply by biemster** [reply]
Thanks for that info, I'm learning a lot on this project :+1:

**Reply by biemster** [reply]
> They are not normally required because many transformers do not have the windings connected. I can't find the reference of the transformer from the model and brand of your switch.

Just for completeness of the conversation, I opened the router and it has H4001CG magnetics.
https://www.digikey.com/en/products/detail/mentech-technology-usa-inc/H4001CG/12683649
![image](https://user-images.githubusercontent.com/5699190/207048240-2c9cf0b6-5009-4e2d-9c80-6f95013e0874.png)

**Reply by holysnippet** [reply]
Hello, okay.
But I don't quite understand how you connect the Pico to Ethernet. You shouldn't connect the Pico directly to the router, it won't work well. You have to follow the diagram and the explanations provided on the documentation.
An additional transformer is required for normal operation.

**Reply by biemster** [reply]
I realize now that your picture with the setup using the magjack got me confused, I was under the impression that the extra transformer would make it more reliable, but was not strictly necessary. I'm trying to fit both the pico and the interface into a very tiny space (PoE adapter), and I don't have room for an extra transformer.
The TX side seems to work without it, is the RX side more picky on this?

**Reply by holysnippet** [reply]
Don't have enough room for a Magjack? That's going to be a problem but there are very small ethernet transformers. You have to find them; or you can wind them by hand if you find the right ferromagnetic material.
If you don't use a transformer for reception then it won't work or will work very little, it will cause problems with the lwIP stack.

This is a bad choice.

**Reply by biemster** [reply]
> Don't have enough room for a Magjack?

I want them to fit in here: https://hackaday.io/project/187979/gallery. I know it might get tricky, but I'm considering this PoE + pico as a base for a lot of other ideas I currently have. And it would save a lot of hassle if that fits in the housing the PoE adapter comes in.
I'll try this: https://hackaday.com/2022/12/03/bit-banging-bidirectional-ethernet-on-a-pi-pico/#comment-6542986
before I start winding my own tiny transformers :smile_cat: , I'll report back here if I manage to get it to work.

**Reply by holysnippet** [reply]
My advice is simple, I think you are wasting your time. The transformer is mandatory for many reasons (that I don't have time to list). Good luck!

**Reply by holysnippet** [reply]
I sent a comment on HaD, if he knows better he could at least explain how to help us! My opinion is that he doesn't even know what he is talking about. He didn't even read the doc before posting.

**Reply by biemster** [reply]
Sorry, I did not mean to offend you. This project is absolutely great, and I will definitely use it often now and in the future. My knowledge of analog electronics is really limited as I said, I'm just trying to get it running with the stuff I have here on my bench. When that fails, and I have to go buy stuff and wait for it 2 months, first thing is of course the transformers.
I'm still learning how this all works, and I find the best way to do that is follow instructions where possible and experiment where needed.
Again, this project and the help is greatly appreciated.

**Reply by holysnippet** [reply]
Don't worry, I don't feel upset or bad. I love your interest in this project.
I strongly disagree with this person's comment. He talks about Gigabit Ethernet and didn't take the time to look at the wiring diagram or the paragraph explaining what is going on with the transformer secondary.
On top of that, he is giving you advice that wastes your time.
It is a lot of work to have developed the electronics, the code and finally the documentation. I am aware that the doc is not clear but I lack time to assemble and synthesize the content.

**Comment by holysnippet** [comment]
Hello, I made a quick correction. I added the initialization of the tx gpios. I thought (and still think) that it is already done correctly by pio_gpio_init, but I added two lines of initialization (just in case).

**Comment by holysnippet** [comment]
Remember also in case you use two Picos simultaneously on the same network that you must not only change the IP address but also the MAC address.

**Comment by holysnippet** [comment]
Hi, I may have one more lead to help you with your diagnosis.
If you use my code in one of your programs, consider setting the Pico's system clock to 120MHz or 240MHz.
It is at the beginning of the main in my program:
set_sys_clock_khz(120000, true);

Good luck !

**Reply by biemster** [reply]
Thanks, I'll add that. I did not have much time to debug this, but I'll update here when there is progress.
